dhis2/dhis2-core
# feat: OpenAPI generator and endpoint [DHIS2-14004]

### Summary

* adds general OpenAPI tool-chain
* adds new OpenAPI controller ("life" analysis)
* adds command line tool ("static" analysis)
* adds `@OpenApi` family of annotations to tweak the generated document in the controllers
* adds a group tag to every controller (via `@OpenApi.Tags`) so they can be split into named groups
* replaces "invisible" parameters for Gist API with a dedicated parameter object
* replaces "invisible" parameters for Datastore query API with a dedicated parameter object
* adds `@OpenApi` annotations to CRUD controller base classes (so inherited methods are mostly described)
* adds some more `@OpenApi` annotations to individual controllers mostly to correct the response type (where it is not visible from the method signature)

Many controllers will need further tweaking by adding `@OpenApi` annotations but the general mechanisms should exist.

### Automatic Testing
The controllers where the API was actually changed by adding dedicated parameter objects is tested by many tests.
New tests were added to test the added OpenAPI document generation endpoints.

### Manual Testing
* make sure controller classes are compiled
* run `OpenApiTool` via `main` method

### Examples
Include all controllers and group them into documents (files) by their first tag (on the controller class):
```bash
--group .

Generated Documents
  openapi-analytics.json         [ 12 controllers, 114 schemas]
  openapi-data.json              [ 29 controllers, 217 schemas]
  openapi-messaging.json         [  7 controllers, 107 schemas]
  openapi-metadata.json          [ 51 controllers, 280 schemas]
  openapi-system.json            [ 20 controllers, 173 schemas]
  openapi-tracker.json           [ 33 controllers, 291 schemas]
  openapi-ui.json                [ 12 controllers, 132 schemas]
  openapi-user.json              [ 14 controllers,  99 schemas]
```

Generate a document just for `/users` path/controller:
```bash
/users openapi-users-only.json

Generated Documents
  openapi-users-only.json        [  1 controllers,  61 schemas]
```

Generate a document with all controllers tagged `"user"`:
```bash
user openapi-user.json

Generated Documents
  openapi-user.json              [ 14 controllers,  99 schemas]
```
